berkeley-function-call-leaderboard/model_handler/storm_handler.py

```diff
@@ -0,0 +1,96 @@
+import re
+import ast
+import inspect
+
+from model_handler.oss_handler import OSSHandler
+from model_handler.utils import convert_to_tool
+from model_handler.constant import GORILLA_TO_OPENAPI
+from model_handler.model_style import ModelStyle
+from model_handler.utils import ast_parse
+
+
+def dict_to_function_format(data):
+    tool_name = data['tool_name']
+    arguments = data.get('tool_arguments', {})
+    args_str = ', '.join(f"{k}={repr(v)}" for k, v in arguments.items())
+    return f"{tool_name}({args_str})"
+
+
+def process_multiple_tool_calls(input_string):
+    # Extract tool calls using regex
+    tool_calls = re.findall(r'<tool_call>(.*?)</tool_call>', input_string, re.DOTALL)
+
+    results = []
+    for call in tool_calls:
+        try:
+            # Convert string representation of dict to actual dict
+            call_dict = ast.literal_eval(call)
+            # Convert dict to function format
+            results.append(dict_to_function_format(call_dict))
+        except (ValueError, SyntaxError) as e:
+            print(f"Error processing tool call: {e}")
+            print(f"Problematic call: {call}")
+
+    return results
+
+
+class StormHandler(OSSHandler):
+    def __init__(self, model_name, temperature=0.001, top_p=1, max_tokens=1000, dtype="bfloat16") -> None:
+        super().__init__(model_name, temperature, top_p, max_tokens, dtype)
+
+    def _format_prompt(prompts, function, test_category):
+        function = convert_to_tool(
+            function, GORILLA_TO_OPENAPI, ModelStyle.OSSMODEL, test_category
+        )
+        tool_call_format = """{"tool_name": <function-name>, "tool_arguments": <args-dict>}"""
+
+        formatted_prompt = inspect.cleandoc(
+            """<|begin_of_text|><|start_header_id|>system<|end_header_id|>
+
+            You are a function calling AI model. You may call one or more functions to assist with the user query. Don't make assumptions about what values to plug into function. The user may use the terms function calling or tool use interchangeably.
+
+            Here are the available functions:
+            <tools>{tools}</tools>
+
+            Follow the below guidelines:
```

Comment:
These guidelines are mentioned in the 'Prompt Format for Function Calling' section in the [model documentation](https://huggingface.co/blog/akjindal53244/llama31-storm8b#prompt-format-for-function-calling). They need to be aligned.

Reply:
`Llama-3.1-Storm-8B` is not trained on negative data. Since model has high instruction following (IFEval) capability, we added two instructions to compensate for the lack of negative data in training. It is also consistent with our evaluation prompt in the blog post: https://huggingface.co/blog/akjindal53244/llama31-storm8b#bfcl

Reply:
Hi @akshita-sukhlecha,

Thank you for the clarification. To ensure consistency and fairness across all models, it's important that the model handler uses the same prompt format as outlined in its official documentation.

In this case, while you have a specialized prompt optimized for BFCL categories, it would be more equitable to use the general-purpose prompt specified in the documentation for a uniform comparison with other models.  

cc @ShishirPatil